### DISCUSSION TITLE
Airplay as Music Provider

### DISCUSSION THREAD

**Discussion by nderooij** [music-assistant/support#1757] [Feature requests and ideas]
### Current Status
As at August 2025 work has commenced on this but is on hold due to other priorities 

### Original Post
Not sure if fully in line with the scope of this project, but just wanted to raise the suggestion :)

How about adding Airplay as player provider? To allow people to easily either select multiple speakers that don't have native Airplay functionality, or Airplay to the group player providers in MA?

Thanks!!

**Comment by OzGav** [comment]
Given that the purpose of MA is to serve your music via a centralised database what would you be airplaying to those speakers?

**Reply by nderooij** [reply]
Good question... It's mostly Spotify and TuneIn, which I can do via MA. My thought was it's easier for my girlfriend to just Airplay, instead if going through another platform she's not used to work with.

**Reply by OzGav** [reply]
OK so what advantage does MA provide for her?

**Reply by nderooij** [reply]
Ease of use, I don’t have to instruct on how MA works, but she can play music throughout the entire house.

But I’m starting to feel like that’s not the intention of MA, that’s fine. I was raising a thought.

**Reply by OzGav** [reply]
If I understand you correctly the general idea has been floated by the devs although not sure if they had Airplay in mind at the time. This would allow MA to do the grouping and sync playback where supported. If your girlfriend isn’t looking to be able to do that then I don’t see any advantage to putting MA in the playback path, you would be better to just stream directly to the player.

**Reply by matejdro** [reply]
My use case for this would be streaming podcasts from my phone to my room music system.

There does not appear to be a good solution at the moment for bridging podcasts into Music Assistants, so just streaming them from my phone seems to be the next best thing.

**Reply by User873902** [reply]
The advantage of using Music Assistant as a conduit over direct streaming would be:

1. AirPlaying to synced heterogeneous speaker groups
2. AirPlaying to speakers non-supported by Apple AirPlay (i.e. Google Speakers)

Music Assistant already has the output part of the puzzle figured out. But, the input is limited to only specific sources.

The success of the AirPlay/Google Cast framework hinges on the ability to spontaneously and effortlessly select desired input from online libraries from a local device, like YouTube, X, or Apple Podcasts.  The content is often single-use and does not need to be a repeatable playlist. It doesn't need to be an automated workload, just an on-the-fly decision. AirPlay is seamlessly integrated into the Apple ecosystem.

Although MA focuses on building a collected library of music content, the same users also enjoy elongated content from YouTube, X, and Apple Podcasts and want to listen to the audio on their home speaker setups.  Most of this content is speech - podcasts, audiobooks, interviews, news, lectures, short-stories, MS Teams meetings, Zoom meetings, etc. Bluetooth only allows playback to one speaker.

Understandably, the developers do not have to choose to make their product a solution for this problem.

**Comment by asosnovsky** [comment]
I second this, while it's nice to be able to use speaker groups inside of MA for streaming for these services. Sometimes I have guests who I would not expect to give access to MA on their phones. It would be nice to expose speakers (or speaker groups) that do not have native support for Airplay. Similarly to how the Aircast HA-Addon does for chromecasts.

**Reply by User873902** [reply]
Thanks for suggesting Aircast! Looks like Aircast requires HA-addons and can't be run directly from a docker container. Otherwise, it would seem an amazing solution for my needs.

**Comment by Juggler00** [comment]
I would also like to see this. I am currently using an LMS instance, and frequently Airplay a podcast from my phone to one of the players. This allows me to keep track of the progress.

Wondering if this project could work. It is already packaged as a Docker container, would just need to find a way to input the audio output into MA: https://github.com/mikebrady/shairport-sync

**Reply by matejdro** [reply]
I'm currently using `shairport-sync` to stream into MPD:

1. Shairport outputs sound into raw file
2. VLC opens this RAW file and creates a HTTP audio stream
3. MPD opens this stream as a HTTP radio stream

This works, but it is very hacky and has considerable latency because of buffering on the MPD side(especially annoying with podcasts, where you pause the podcast and it keeps playing for a second or two).

I asked whether MA can be configured to be better in this regard, but no replies yet: https://github.com/orgs/music-assistant/discussions/2253

**Reply by sbelyea** [reply]
> I'm currently using `shairport-sync` to stream into MPD:
> 
> 1. Shairport outputs sound into raw file
> 2. VLC opens this RAW file and creates a HTTP audio stream
> 3. MPD opens this stream as a HTTP radio stream
> 
> This works, but it is very hacky and has considerable latency because of buffering on the MPD side(especially annoying with podcasts, where you pause the podcast and it keeps playing for a second or two).
> 
> I asked whether MA can be configured to be better in this regard, but no replies yet: https://github.com/orgs/music-assistant/discussions/2253

@matejdro  - I'm actually very interested in the specifics of how you're achieving this.  Moving my squeezebox clients from a dedicted Lyrion server to Music Assistant has the unfortunate side effect of _not_ being able to AirPlay to those speakers.  Sounds like you manufactured a workaround to still have these Squeezebox devices show up as AirPlay targets with Shairport-Sync.  Curious about how you're doing the output from Shairport, streaming via VLC and MPD radio stream aspects.

**Reply by matejdro** [reply]
Here you go, I have written a little writeup: https://gist.github.com/matejdro/b902709d10d0c86e3b68a2b00102f438

**Comment by firefrei** [comment]
I would also love to see airplay as music provider!

This is my setup:  
I am streaming to multiple rooms using snapcast and MA. This works very well. In addition, I have added shairport-sync as a source on my snapcast server - and control speaker settings, volume levels, etc. via Home Assistant automations and the snapcast integration.
I am facing the challenge that MA creates a new, temporary snapcast source, and groups snapclient players once something is played. When I want to stream via Airplay now, I manually need to adjust the snapcast configuration outside MA. So double configuration overhead, partly-conflicting setups, and a less intuitive setup to control for the non-technicians in my home :-)

From my perspective, it would be great to have MA as the central entity for music/audio management of my home. This would also create a clear and linear call flow, and defined responsibilities:
Control: HA -> MA -> Snapserver
Streaming: Airplay/Spotfiy/Radio/Soundcloud/Files/... -> MA -> Snapclients

**Reply by firefrei** [reply]
Alternatively, a generic music provider could also help here. For example, a pipe input like snapcast provides. So I could add shairport as a music provider over the pipe provider, and control all the audio routing with MA.

#1114

**Comment by Juggler00** [comment]
Any chance the devs are looking at this?

**Comment by iko3** [comment]
I also have a use case for AirPlay as a music source. I have several speaker zones in my house, which I connected to my home server where I run multiple instances of shairport-sync to be able to stream to each of them individually. 
I would rather have a centralized solution that allows me to both AirPlay or play directly from Spotify or Internet radio.

**Reply by iVolt1** [reply]
I made AirPlay a music source in the HomePod Connect addon by adding an instance of shaiprort-sync that outputs to a the same pipe that librespot-java outputs to in the addon.  The addon also has code to make OwnTone ignore that AirPlay instance to avoid loping back and has code to stop play from librespot-java to prvent both inputs writing to the pipe.

It is a great way to allow any device that can output to AirPlay a source for multiroom audio with OwnTone and control the outputs from the HA dashboard.  There is a little lag on starting and stopping play.  Otherwise it works great.

Having this capability in Music Assistant could be very helpful for lots of users.

**Comment by GentleHoneyLover** [comment]
+1

**Comment by mr-ransel** [comment]
Adding another general usecase for this: The ability to support playing whole-home audio from sources that do not implement a way to expose the libraries, but play the audio locally on your device. 

What:
There are hundreds of examples, but say [PocketBard](https://apps.apple.com/us/app/pocket-bard-app/id6444221555) for example, for playing D&D ambience music, or any number of sound generator or meditation apps on the market. None of them will ever be worthwhile implemented as Providers separately, but are completely reasonable inputs into a whole-home audio system.

Why:
Music Assistant is a fantastic 1-stop shop for turning the hodgepodge of components I have around my house into a whole home music system, but it lacks all of the general "I have something else" input support methods (Local audio provider, Airplay provider, ChromeCast Audio provider). The only exception to this I've seen is if you can jump through the hoops of setting up icecast, or another HTTP stream for a radio station, and even that renders all mobile device outputs unavailable for this purpose unless you build yet another component that implements an airplay receiver and pipes it into icecast on your behalf.

All of the things too small, or that the developer team isn't staffed to implement would be able to leverage this escape hatch and benefit from being able to play across all the different devices MA supports for playback.

**Comment by kringsc** [comment]
Adding to this request. Anybody found a good way to do this?

**Comment by prclm** [comment]
+1

**Reply by adamcstephens** [reply]
Please use the reactions at the top of the discussion. A comment of +1 pings everyone watching this without adding any value.

**Comment by anthonyfilmsstuff** [comment]
Yes, I would like to express my desire to have this feature added. I migrated to Music Assistant from LMS a few months back, and honestly this is the only feature I miss from LMS. Having the ability to stream to any, some, or all of my speakers via Airplay was great. It was also an excellent and simple way for guests or my wife to pull up a song or playlist they had on their phone to listen to it without having to touch the LMS interface and just use what they're used to.

**Comment by OzGav** [comment]
This has been added to [the latest nightly](https://github.com/music-assistant/server/releases/tag/2.7.0.dev2025110701)

**Reply by Juggler00** [reply]
Amazing!!! Will be watching the releases closely!!!

**Reply by marcelveldt** [reply]
I think we can do another beta today or tomorrow so you can test it out

**Reply by scyto** [reply]
i was about to add a feature request to let me expose one per squeezelite player i have, then realized i was dumb*** and can add multiple and rename and target :-) in the same way you made me realize how to add multiple providers :-)

it would be nice if this drop down showed the player icon type..... yes i have multiple players in the same room using different player protcols, obvioulsy i could fix using a prefix.... but still would be nice

or add 'enable airplay' to the advanced settings page of each of the other player types ;-). (i can't help myself i am a prodcut manager by day and my devs hate me)

<img width="1196" height="781" alt="image" src="https://github.com/user-attachments/assets/2d05313e-2ae3-43c7-b348-63e2fe858fef" />

**Reply by scyto** [reply]
I added 8 airplay providers to match by 8 squeezelite players - why does the (AirPlay) after the player name come and go in the interface for all those?

<img width="1251" height="677" alt="image" src="https://github.com/user-attachments/assets/c4135093-6f73-42e3-9d6c-1ea759df9fff" />

they are all configured like this....

<img width="1279" height="661" alt="image" src="https://github.com/user-attachments/assets/5b9e2037-a38b-469b-a639-69c7aa28ddbc" />


<img width="1259" height="811" alt="image" src="https://github.com/user-attachments/assets/5eeffdfd-637b-4e66-b065-e555bae3e42e" />

**Reply by OzGav** [reply]
Ordinarily, the suffix is for players that have another primary type. Are all of these the same device?

**Reply by scyto** [reply]
> Ordinarily, the suffix is for players that have another primary type. Are all of these the same device?

They are all squeezelite instances running on the same hardware, each instance has its own mac address.  
Which devices get (AirPlay) after them seemed to rotate as a i added each airplay sink

i notice this morning (AirPlay) is removed from all of them, here you can see the music assistant airplay devices i created (all with _) and most of the squeezelite instances they service

one thesis - maybe the (AirPlay) gets added when first created, then i have to go back in and maned name with _ (i am trying to figure out how to make the _ ones appear first in my airplay list) and then maybe because it has a _ and so name is no longer a dupe it removes the (AirPlay) after some time. - if so i think thats a ok 'as designed' issues, was just confusing

<img width="699" height="1458" alt="image" src="https://github.com/user-attachments/assets/cdbddc7f-9a95-4aaf-8bf7-e5b521c1a96c" />

**Reply by scyto** [reply]
also not sure it needs (AirPlay) after it as it says it - also doesn't seem to happen for say a Sonos. These are the same physical device. And this is the AirPlay the Sonos is advertising itself

<img width="700" height="127" alt="image" src="https://github.com/user-attachments/assets/ec779e04-e0ca-4bd0-ac5f-58368cb6b37c" />

so no (AirPlay) added there either?

however for my Sony Amp, it did get added this is the Sony advertising its owne airplay sink and  DLNA sink 

<img width="693" height="129" alt="image" src="https://github.com/user-attachments/assets/55f6ac53-f88e-4b9a-b671-f626da32ee1d" />

i think i understand what you are trying to do - differentiate what player type identically named players are?

if so for this UI adding an icon would be great next to the player name (DLNA, squeezelite, airplay)

<img width="457" height="922" alt="image" src="https://github.com/user-attachments/assets/fa2f352a-4b08-45d8-86e7-f5e8d105d97d" />

so in this example the Dendon would get an airplay logo next to one instance and a DLNA logo next to the other instance....

of course that doesn't help with programmtic access from HA..... i need to think about that.....

same here would a logo work?

<img width="349" height="932" alt="image" src="https://github.com/user-attachments/assets/4073786e-e82b-4293-b8b7-6feb9d61f12c" />

tbh maybe just a global option that adds `- source type` after all names - both for UI and programmatic acccess, could be off by default for people with simple setups and to reduce annoyance of having the type added after every name?

(also why does _ not work in this UI to push the _ items to top - thats the usual pattern for most software, is there another char i should use?, is there a way for explicit order?)

**Comment by OzGav** [comment]
The feature request has been implemented successfully. As such this feature request will be closed and locked. If there are enhancements desired then open a new [feature request](https://github.com/orgs/music-assistant/discussions/categories/feature-requests-and-ideas). If there is a question about how this works then check the [documentation](https://beta.music-assistant.io/), or ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord. If there are problems then ask a question in the [Q&A](https://github.com/orgs/music-assistant/discussions/categories/q-a) area or on Discord or open an [issue](https://github.com/music-assistant/support/issues). Thanks for your interest in seeing this idea being brought to life.
